afpacket/tests/afpacket_tests.rs

```diff
@@ -0,0 +1,87 @@
+#![cfg(target_os = "linux")]
+
+use afpacket;
+use rand::{self, Rng};
+use smoltcp::{phy::ChecksumCapabilities, wire};
+use std::{ffi::CString, sync::mpsc, thread, time::Duration};
+
+#[test]
+fn layer2_loopback() {
+    // If this takes more than a second to occur, something's definitely wrong.
+    let timeout = Duration::from_secs(1);
+
+    let mut rng = rand::thread_rng();
+
+    let iface_name = CString::new("lo").unwrap();
+
+    let side_a = afpacket::Socket::new().unwrap();
+    let mut side_a = side_a.bind(&iface_name).unwrap();
+
+    let side_b = afpacket::Socket::new().unwrap();
+
+    let (tx, rx) = mpsc::channel();
+
+    let thread_b = thread::spawn(move || {
+        let mut side_b = side_b.bind(&iface_name).unwrap();
+        side_b.set_promisc().unwrap();
+
+        println!("b: recving packet");
+        let mut in_buffer = vec![0; 1500];
+        let (len, _) = side_b.recv(&mut in_buffer).unwrap();
+        in_buffer.resize(len, 0);
+        println!("b: recved packet");
+
+        tx.send(in_buffer).unwrap();
+    });
+
+    // now send a packet from side a to side b
+    let body = {
+        let mut body = vec![0; 64];
+        rng.fill(&mut body[..]);
+        body
+    };
+    let udp_hdr = wire::UdpRepr {
+        src_port: 3001,
+        dst_port: 3002,
+        payload: &body,
+    };
+    let src_addr = wire::Ipv4Address::new(10, 0, 0, 1);
+    let dst_addr = wire::Ipv4Address::new(10, 0, 0, 2);
+    let ipv4_hdr = wire::Ipv4Repr {
+        src_addr,
+        dst_addr,
+        protocol: wire::IpProtocol::Udp,
+        payload_len: udp_hdr.buffer_len(),
+        hop_limit: 2,
+    };
+    let eth_hdr = wire::EthernetRepr {
+        src_addr: wire::EthernetAddress::BROADCAST,
+        dst_addr: wire::EthernetAddress::BROADCAST,
+        ethertype: wire::EthernetProtocol::Ipv4,
+    };
+    let buf_size =
+        wire::EthernetFrame::<&[u8]>::buffer_len(ipv4_hdr.buffer_len() + udp_hdr.buffer_len());
+    let mut out_buffer = vec![0; buf_size];
+    {
```

Comment:
What's with the extra block scope?

Reply:
Forgot that's now legal with NLL. 
(My concern was with the mutable borrow followed by the immutable borrow down below.)